File: sdk/python/polar/event_types.py
from __future__ import annotations

import builtins
import typing

from polar.base import AsyncServiceBase, SyncServiceBase, parse_response
from polar.errors import (
    HTTPValidationError,
    Update404Error,
)
from polar.inputs import (
    EventTypeUpdate,
)
from polar.literals import (
    EventSource,
    EventTypesSortProperty,
)
from polar.outputs import (
    EventType,
    ListResourceEventTypeWithStats,
)


class EventTypesSync(SyncServiceBase):
    def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        root_events: bool = False,
        parent_id: str | None = None,
        source: EventSource | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[EventTypesSortProperty] | None = ["-last_seen"],
    ) -> ListResourceEventTypeWithStats:
        """
        List event types with aggregated statistics.

        **Scopes**: `events:read` `events:write`

        Args:
            organization_id: Filter by organization ID.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by external customer ID.
            query: Query to filter event types by name or label.
            root_events: When true, only return event types with root events (parent_id IS NULL).
            parent_id: Filter by specific parent event ID.
            source: Filter by event source (system or user).
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/event-types/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "query": query,
                "root_events": root_events,
                "parent_id": parent_id,
                "source": source,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceEventTypeWithStats, method_errors)

    def update(
        self,
        id: str,
        **kwargs: typing.Unpack[EventTypeUpdate],
    ) -> EventType:
        """
        Update an event type's label.

        **Scopes**: `events:write`

        Args:
            id: The event type ID.

        Raises:
            Update404Error: Not Found
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/event-types/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = self.client.send_request(request)
        method_errors = {
            404: Update404Error,
            422: HTTPValidationError,
        }
        return parse_response(response, EventType, method_errors)


class EventTypesAsync(AsyncServiceBase):
    async def list(
        self,
        *,
        organization_id: str | builtins.list[str] | None = None,
        customer_id: str | builtins.list[str] | None = None,
        external_customer_id: str | builtins.list[str] | None = None,
        query: str | None = None,
        root_events: bool = False,
        parent_id: str | None = None,
        source: EventSource | None = None,
        page: int = 1,
        limit: int = 10,
        sorting: builtins.list[EventTypesSortProperty] | None = ["-last_seen"],
    ) -> ListResourceEventTypeWithStats:
        """
        List event types with aggregated statistics.

        **Scopes**: `events:read` `events:write`

        Args:
            organization_id: Filter by organization ID.
            customer_id: Filter by customer ID.
            external_customer_id: Filter by external customer ID.
            query: Query to filter event types by name or label.
            root_events: When true, only return event types with root events (parent_id IS NULL).
            parent_id: Filter by specific parent event ID.
            source: Filter by event source (system or user).
            page: Page number, defaults to 1.
            limit: Size of a page, defaults to 10. Maximum is 100.
            sorting: Sorting criterion. Several criteria can be used simultaneously and will be applied in order. Add a minus sign `-` before the criteria name to sort by descending order.

        Raises:
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="GET",
            url="/v1/event-types/",
            path_params={},
            query_params={
                "organization_id": organization_id,
                "customer_id": customer_id,
                "external_customer_id": external_customer_id,
                "query": query,
                "root_events": root_events,
                "parent_id": parent_id,
                "source": source,
                "page": page,
                "limit": limit,
                "sorting": sorting,
            },
        )
        response = await self.client.send_request(request)
        method_errors = {
            422: HTTPValidationError,
        }
        return parse_response(response, ListResourceEventTypeWithStats, method_errors)

    async def update(
        self,
        id: str,
        **kwargs: typing.Unpack[EventTypeUpdate],
    ) -> EventType:
        """
        Update an event type's label.

        **Scopes**: `events:write`

        Args:
            id: The event type ID.

        Raises:
            Update404Error: Not Found
            HTTPValidationError: Validation Error
        """
        request = self.client.build_request(
            method="PATCH",
            url="/v1/event-types/{id}",
            path_params={
                "id": id,
            },
            query_params={},
            body=kwargs,
        )
        response = await self.client.send_request(request)
        method_errors = {
            404: Update404Error,
            422: HTTPValidationError,
        }
        return parse_response(response, EventType, method_errors)

File: sdk/typescript/src/2026-10/services/event_types.ts
import type { ClientBase, RequestOptions } from "../../base";
import type {
  EventSource,
  EventType,
  EventTypeUpdate,
  EventTypeWithStats,
  EventTypesSortProperty,
  ListResourceEventTypeWithStats,
} from "../models";

import { EventTypesUpdate404Error, HTTPValidationError } from "../errors";

export const listEventTypes = (client: ClientBase) => {
  /**
   * List event types with aggregated statistics.
   *
   * **Scopes**: `events:read` `events:write`
   *
   * @param query - Query parameters
   * @param requestOptions - Request options
   * @returns {ListResourceEventTypeWithStats}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    query?: {
      organization_id?: string | string[] | null;
      customer_id?: string | string[] | null;
      external_customer_id?: string | string[] | null;
      query?: string | null;
      root_events?: boolean;
      parent_id?: string | null;
      source?: EventSource | null;
      page?: number;
      limit?: number;
      sorting?: EventTypesSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): Promise<ListResourceEventTypeWithStats> => {
    const pathParams = {};
    const queryParams = {
      organization_id: query?.organization_id,
      customer_id: query?.customer_id,
      external_customer_id: query?.external_customer_id,
      query: query?.query,
      root_events: query?.root_events,
      parent_id: query?.parent_id,
      source: query?.source,
      page: query?.page ?? 1,
      limit: query?.limit ?? 10,
      sorting: query?.sorting ?? ["-last_seen"],
    };
    const request = client.buildRequest(
      "GET",
      "/v1/event-types/",
      pathParams,
      queryParams,
      undefined,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<ListResourceEventTypeWithStats>(response, "json", {
      422: HTTPValidationError,
    });
  };
};
/**
 * List event types with aggregated statistics.
 *
 * **Scopes**: `events:read` `events:write`
 *
 * @param query - Query parameters
 * @param requestOptions - Request options
 * @returns {AsyncGenerator<EventTypeWithStats>} A generator that yields items of type EventTypeWithStats.
 * @throws {PolarNetworkError} When a network error occurs
 * @throws {PolarRateLimitError} When the rate limit is exceeded
 * @throws {PolarServerError} When the server returns a 5xx error
 * @throws {HTTPValidationError} Validation Error
 */
export const iterListEventTypes = (client: ClientBase) => {
  return async function* (
    query?: {
      organization_id?: string | string[] | null;
      customer_id?: string | string[] | null;
      external_customer_id?: string | string[] | null;
      query?: string | null;
      root_events?: boolean;
      parent_id?: string | null;
      source?: EventSource | null;
      page?: number;
      limit?: number;
      sorting?: EventTypesSortProperty[] | null;
    },
    requestOptions?: RequestOptions,
  ): AsyncGenerator<EventTypeWithStats> {
    let page: number;
    page = query?.page ?? 1;
    let limit: number | undefined;
    limit = query?.limit;

    while (true) {
      const response = await listEventTypes(client)({ ...query, page, limit }, requestOptions);
      for (const item of response.items) {
        yield item;
      }
      if (page >= response.pagination.max_page) {
        break;
      }
      page++;
    }
  };
};
export const updateEventTypes = (client: ClientBase) => {
  /**
   * Update an event type's label.
   *
   * **Scopes**: `events:write`
   *
   * @param id - The event type ID.
   * @param body - Request body
   * @param requestOptions - Request options
   * @returns {EventType}
   * @throws {PolarNetworkError} When a network error occurs
   * @throws {PolarRateLimitError} When the rate limit is exceeded
   * @throws {PolarServerError} When the server returns a 5xx error
   * @throws {EventTypesUpdate404Error} Not Found
   * @throws {HTTPValidationError} Validation Error
   */
  return async (
    id: string,
    body: EventTypeUpdate,
    requestOptions?: RequestOptions,
  ): Promise<EventType> => {
    const pathParams = {
      id: id,
    };
    const queryParams = {};
    const request = client.buildRequest(
      "PATCH",
      "/v1/event-types/{id}",
      pathParams,
      queryParams,
      body,
    );
    const response = await client.sendRequest(request, requestOptions);
    return client.parseResponse<EventType>(response, "json", {
      404: EventTypesUpdate404Error,
      422: HTTPValidationError,
    });
  };
};

export function createEventTypesService(client: ClientBase) {
  return {
    list: listEventTypes(client),
    update: updateEventTypes(client),
    iterList: iterListEventTypes(client),
  };
}

export type EventTypes = ReturnType<typeof createEventTypesService>;
